### PR TITLE
Drop `SafeRefCount` in favor of C++11 atomics

### LINKED ISSUE #31860: Atomic operations in SefRefCount are not properly cross-platform
**Godot version:**
3.2 master (adb85fc8)

**OS/device including version:**
All, but known problems on OpenBSD/macppc so far, or Android/x86 if `__atomic*` methods are used.

**Issue description:**
#31321 by @julianaito aimed to address a portability issue for `SefRefCount`'s atomic operations on OpenBSD/macppc, but caused an issue with Android/x86 (#31407).
It was thus reverted for the time being, but a better solution needs to be looked into and merged again.

See discussion in #31407, and the possibility to simply drop `SafeRefCount` in 4.0 to use atomic types from the standard library instead.

CC @RandomShaper @hpvb @reduz

### PR DESCRIPTION
**Pull request by RandomShaper**
Primitive variables being treated as atomics are also promoted to `std::atomic<T>`.

`atomic_conditional_increment` and `atomic_exchange_if_greater` still exist since they are useful, but are also meant to be used on C++11 atomics.

`CowData` specifics:
- There's an `static_assert` near the top, ensuring that the memory layout of a `std::atomic<uint32_t>`  is just the `uint32_t` itself, or that at least it doesn't require more memory than it.
  On the other hand, the standard guarantees that integer atomics are trivially copyable so it's OK to use raw memory functions on them (and `reinterpret_cast`), just as if it were a plain `uint32_t`.
  Anyway, when initializing the refcount it's made explicit that the intent is that they are atomic integers, even though the resulting machine code would be the same.
- Removed misleading parameter from `_unref`.

Fixes #31860.

@akien-mga, I'm letting you set the milestone. Probably this is OK for 3.2 (I think none of the removed items were exposed to scripting of GDNative, but I'm not sure). Otherwise, or just for some safety, this should be for 4.0.

### CODE REVIEW

**Comment by Calinou** [comment]
Out of curiosity, does this have performance benefits?

**Comment by RandomShaper** [comment]
I don't think so. Unless the old code was invoking some lock-based implementation on some platform or something, but doesn't look like it.

File: core/array.cpp [hunk 1]
```diff
@@ -34,10 +34,11 @@
 #include "core/object.h"
 #include "core/variant.h"
 #include "core/vector.h"
+#include <atomic>
```
Comment (akien-mga) [inline core/array.cpp]: Style nitpick, but there should be an empty line between our headers `"..."` and system headers `<...>`.
http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

**Comment by akien-mga** [comment]
Given the timing, I'd opt for cautiousness and target this for 4.0. I really want to get 3.2 released ASAP :)

Moreover I think these changes may partly conflict with changes done in the `vulkan` branch by @reduz, so merging them for 3.2 would make rebasing `vulkan` trickier.

That being said, if @hpvb or @reduz want these changes done for 3.2, I'm not against it.

File: core/array.cpp (hunk 1 above)
Comment (RandomShaper) [reply core/array.cpp]: Fixed.

**Comment by RandomShaper** [comment]
I've tried changing the base branch to `vulkan`, but the CI still believes this is unmergeable, so I'm closing this and will open a new one directly with the new base.

**Pull request closed without merging**